### DISCUSSION TITLE
Decompressing meshes in bulk / merging (CLI)

### DISCUSSION THREAD

**Discussion by ArchieLuxton** [donmccurdy/glTF-Transform#1086] [Q&A] [answered]
Hi,

I've been experimenting with the fastest way to decompress draco-compressed meshes (anywhere between 1 and 30 at a time) and would really appreciate some pointers. 

At the moment my app is all in Python, and is calling the CLI in parallel to `copy` the compressed `.gltf` files into decompressed `.glb` files. For around 15 meshes this is taking ~4s, but I'm really keen to reduce this as much as possible. At the moment I'm feeding in multiple meshes to my app, but it could just as easily be one big mesh. 

I've experimented with the `merge` functionality of the CLI, but I'm struggling to get it to work as I expect.

I expect if I merge all my meshes together first and then decompress the combined mesh, this will likely be much faster than my current approach, but is this possible with `merge`? The file size increases and I can see that there are multiple meshes/scenes in the file, but I can only ever see the first mesh when viewing it online (https://gltf-viewer.donmccurdy.com/). 

Any advice would be hugely appreciated on how to decompress these meshes in bulk. Thank you!

**Comment by donmccurdy** [comment]
glTF Transform cannot merge scenes without first decompressing them, so I doubt that will help you with performance. You only see the first input file in the viewer because `merge` creates a glTF file with multiple scenes, all the data is still there, but since most glTF files contain only one scene, most viewers just show that. There are ways around the "multiple scenes" issue but I don't think this is the answer for your performance question, let's look for a better way.

Do you definitely need decompression to happen via CLI called from Python, on the backend? Are you able to share more about what the app is doing, when, and where the data is coming from or going? There could be more efficient ways to structure this.

EDIT: If it's an option - Meshopt compression will decompress much more quickly than Draco compression.

**Comment by ArchieLuxton** [comment]
Hi Don, thanks for the quick response. 

My app is getting draco compressed gltf files (3d tiles) from an API as bytes. It's then temporarily saving them in order to be able to use the CLI to decompress, then loads these into Open3D and deletes the meshes from disk (not ideal, really I'd prefer this to all work without writing to disk at all). 

The app itself performs various geospatial analysis, but for various reasons unfortunately the bulk of this analysis needs to remain in Python/Open3D. The output of the analysis is just coordinates of some vertices of interest, plus other arrays for scalar properties of these vertices. I.e. I don't need to do any exporting or visualisation of meshes or geometry. Nor do I need the textures of the geometry for anything. 

For every location analysed there's around 10-20 tiles/meshes to load in, and this needs to be done on the fly. So really any solution that can quickly process these files in the backend to then pass to Python would be fantastic. Since I don't need textures etc, I could also just pass the triangles and vertices arrays to Python from the backend instead of loading a .glb directly in Open3d if that is simpler?

Thanks again for your assistance.

**Reply by donmccurdy** [reply]
Hm I see, thanks.  So this is 10-20 tiles, and one glTF file per tile? It might help to know what's in the files in more detail, could you share the output of...

```bash
gltf-transform inspect scene.glb --format md
```

... for 1-2 of these files? I want to confirm that each file's data is mostly geometry, and a small number of meshes per file. Or if it's possible to share the files, that's even better.

Assuming my guesses above are true, I don't know of much we can do to improve the decoding speed of any particular file in isolation, but maybe the best option would be to help find a way for you to decode those 10-20 tiles in parallel.

**Reply by ArchieLuxton** [reply]
Yes that's right, there are 10-20 different compresesd gltf files (each downloaded separately from an API). 

Here's the output of a single gltf (which is a single tile). Each of the files are *roughly* the same size.

``
OVERVIEW
 ────────────────────────────────────────────
| key                | value                                           |
| ---                | ---                                             |
| version            | 2.0                                             |
| generator          | draco_decoder                                   |
| extensionsUsed     | KHR_draco_mesh_compression, KHR_materials_unlit |
| extensionsRequired | KHR_draco_mesh_compression, KHR_materials_unlit |



 SCENES
 ────────────────────────────────────────────
| #   | name | rootName | bboxMin                                  | bboxMax                                  |
| --- | ---  | ---      | ---                                      | ---                                      |
| 0   |      |          | 3980794.14659, 4966706.05644, 2433.12201 | 3980822.41756, 4966728.52343, 2481.05158 |



 MESHES
 ────────────────────────────────────────────
| #   | name | mode      | primitives | glPrimitives | vertices | indices | attributes                   | instances | size¹    |
| --- | ---  | ---       | ---        | ---          | ---      | ---     | ---                          | ---       | ---      |
| 0   |      | TRIANGLES | 1          | 978          | 1,305    | u16     | POSITION:f32, TEXCOORD_0:f32 | 1         | 31.97 KB |

¹ size estimates GPU memory required by a mesh, in isolation. If accessors are
  shared by other mesh primitives, but the meshes themselves are not reused, then
  the sum of all mesh sizes will overestimate the asset's total size. See "dedup".



 MATERIALS
 ────────────────────────────────────────────
| #   | name | instances | textures         | alphaMode | doubleSided |
| --- | ---  | ---       | ---              | ---       | ---         |
| 0   |      | 1         | baseColorTexture | OPAQUE    |             |



 TEXTURES
 ────────────────────────────────────────────
| #   | name | uri | slots            | instances | mimeType   | compression | resolution | size    | gpuSize¹  |
| --- | ---  | --- | ---              | ---       | ---        | ---         | ---        | ---     | ---       |
| 0   |      |     | baseColorTexture | 1         | image/jpeg |             | 256x256    | 4.62 KB | 349.52 KB |

¹ gpuSize estimates minimum VRAM memory allocation. Older devices may require
  additional memory for GPU compression formats.
``

Also attached as a picture as it seems the formatting here is off. 

Currently I'm running the CLI in parallel (`concurrent.futures.ThreadPoolExecutor` in Python) but any faster way would be fantastic. Happy to call some JS under the hood. 

![mesh_1](https://github.com/donmccurdy/glTF-Transform/assets/78753319/d58fa64a-f7fc-4a79-a956-67d3dcb17f6d)

**Reply by donmccurdy** [reply]
Hm, decoding 10–20 meshes with 30kb of geometry shouldn't take 4 seconds, especially in parallel.. My guess would be that there's some overhead both in Python starting thread pools and shell contexts, and in glTF Transform initializing the CLI and a Draco WebAssembly decoder each time.

What do you think about doing something like this?

1. Create a minimal node.js script that creates a single NodeIO instance, reads all N tiles, decodes them, and writes to disk.
2. Python invokes that script once, rather than N times.

Node.js has its own worker threading system, which we could include if this seems like a better approach, but I'd try it all in one thread to start.

It might also be worth testing having Python invoke a bash script 10-20 times that does _nothing_, like `sleep 0.25`, just to get a sense of what the overhead is on that side.

**Reply by ArchieLuxton** [reply]
Thanks Don, I'll give that a shot. 

I don't suppose you'd be able to point me towards some example JS code that loads, decodes and writes the files? Unfortunately I don't have much experience with the full gltf-transform or node.js, and I'm having some problems with extensions (e.g. `KHR_materials_unlit`).

Thank you very much for your assistance, it's really appreciated.

**Reply by donmccurdy** [reply]
Sure — here's a quick example.

```javascript
import { NodeIO } from '@gltf-transform/core';
import { KHRONOS_EXTENSIONS } from '@gltf-transform/extensions';
import draco3d from 'draco3dgltf';

const io = new NodeIO()
	.registerExtensions(KHRONOS_EXTENSIONS)
	.registerDependencies({
		'draco3d.decoder': await draco3d.createDecoderModule(),
	});

console.time('decode');

const pending = (new Array(10).fill(0)).map(async (_, index) => {
	const srcPath = Math.random() > .5 ? './tile_1.glb' : './tile_2.glb';
	const document = await io.read(srcPath);

	for (const extension of document.getRoot().listExtensionsUsed()) {
		if (extension.extensionName === 'KHR_draco_mesh_compression') {
			extension.dispose();
		}
	}

	await io.write(`./out/tile_${index}.glb`, document);
});

await Promise.all(pending);

console.timeEnd('decode');
```

I tested with two GLBs from 3D Tiles, one 22kb and the other 95kb. The script above takes 95–100ms to decode 10 tiles on my Macbook, without any threading involved. The performance could probably be improved with threading, but there's some cost to initializing node.js and WASM too, so I'm not sure how much. You could perhaps keep a Node.js server running to handle decoding without re-initializing for each request from Python, but that starts to become more complex to maintain perhaps.

**Reply by ArchieLuxton** [reply]
That works perfectly, I'm getting between 100ms-250ms depending on the OS now for >30 meshes. 

One final question (and a long shot) - is it possible to decode a mesh directly from bytes? I.e. the response from the API. In that instance I suppose it would be ideal to decode it and then return the vertices and triangle arrays to be picked up at the next step. 

Thanks again for your help.

**Reply by donmccurdy** [reply]
Definitely! I'm not familiar with the 3D Tiles API, but in Node.js it should look something like:

```js
import { NodeIO } from '@gltf-transform/core';
import { KHRONOS_EXTENSIONS } from '@gltf-transform/extensions';
import draco3d from 'draco3dgltf';

const url = 'https://tile.googleapis.com/v1/3dtiles/datasets/CgA/files/UlRPVEYuYnVs.json?session=YOUR_SESSION&key=YOUR_API_KEY';

const io = new NodeIO()
	.registerExtensions(KHRONOS_EXTENSIONS)
	.registerDependencies({
		'draco3d.decoder': await draco3d.createDecoderModule(),
	});

const buffer = await fetch(url).then((res) => res.arrayBuffer());
const document = await io.readBinary(new Uint8Array(buffer));

for (const extension of document.getRoot().listExtensionsUsed()) {
	if (extension.extensionName === 'KHR_draco_mesh_compression') {
		extension.dispose();
	}
}

const prim = document.getRoot()
  .listMeshes()[0]
  .listPrimitives()[0];

const vertices = prim.getAttribute('POSITION').getArray();
const triangles = prim.getIndices().getArray();
```

**Reply by donmccurdy** [reply]
I think the time required for the CLI to initialize could be reduced too, but it will never be as quick as a script importing only the modules required for decoding. I've filed an issue about that: 

- https://github.com/donmccurdy/glTF-Transform/issues/1089.

**Reply by ArchieLuxton** [reply]
Thanks again, I've modified the code slightly to save the arrays as csv files for now to rebuild in Open3d and inspect. 

I'm getting some odd artefacts that look like some triangles are missing however - the below is a screenshot of a mesh reconstructed from the two arrays:

![image](https://github.com/donmccurdy/glTF-Transform/assets/78753319/ebfb2f9a-8adc-4037-b331-69ddfbe52358)

And here is the gltf opened in your web viewer:

![image](https://github.com/donmccurdy/glTF-Transform/assets/78753319/5f8c30d1-dc4c-474c-a71c-b9fe0b416747)

Any ideas what could be causing that?

The correct version of the decoded mesh has 6686 triangles and 9161 vertices, and the other version has 5739 triangles and 8381 vertices.

[files.zip](https://github.com/donmccurdy/glTF-Transform/files/12601585/files.zip)

Attached is a zip with the encoded mesh and the exported arrays of triangles and vertices (of shape (N, )) if that helps.

**Reply by donmccurdy** [reply]
One thing i'd missed originally – the tiles all appear to have a single mesh, but that mesh could have multiple primitives. So the extraction should look more like:

```js
for (const mesh of document.getRoot().listMeshes()) {
  for (const prim of mesh.listPrimitives()) {
    const vertices = prim.getAttribute('POSITION').getArray();
    const triangles = prim.getIndices().getArray();
  }
}
```

If we were trying to support any arbitrary glTF input, we'd also need to think about the location of the nodes to which these meshes are attached, but I don't know if that's a concern here.

**Reply by ArchieLuxton** [reply]
Thanks Don, I've now got a script running that can take in the URIs and send back the geometry to my Python process. Working well!

Only a final issue - it appears all the meshes are centred about the same point (when the vertices should relate to the original ECEF coordinate system), is there a simple way of getting the correct offsets from each?

E.g. here's all the meshes for a given search (around 40 here) shown at once

![image](https://github.com/donmccurdy/glTF-Transform/assets/78753319/030684e8-fdd9-4d65-9858-0789aa0bb1fa)

Where it should be something like this (minus the texture):

![image](https://github.com/donmccurdy/glTF-Transform/assets/78753319/8cb93b62-e0d5-4ec5-878f-35ad256ef863)

The vertices for a single mesh (per the output of the latest script) are currently something like:

```
array([[-23.49495888,  21.97915649, -47.74781799],
       [-24.63181305,  22.35810471, -47.74781799],
       [-23.11600876,  21.97915649, -46.61096573],
       ...,
       [  1.51580429,  -1.13685226,  42.82145691],
       [  1.51580429,  -1.13685226,  42.82145691],
       [  3.03160858,  -2.27370644,  47.74781799]])
```

**Reply by donmccurdy** [reply, marked as answer]
The [Scene](https://gltf-transform.dev/modules/core/classes/Scene) and [Node](https://gltf-transform.dev/modules/core/classes/Node) documentation are probably the resources you'll want here. In this case each tile has only a single node, and that node has only a  position, so you can extract mesh position with:

```js
const scene = document.getRoot().listScenes()[0];

scene.traverse((node) => {
  const translation = node.getWorldTranslation();
  for (const prim of node.getMesh().listPrimitives()) {
    const vertices = prim.getAttribute('POSITION').getArray();
    const triangles = prim.getIndices().getArray();
  }
});
```

In other glTF sources you'd want to consider world translation, rotation, and scale, or the world matrix, to get all possible transforms applied to the mesh.

**Comment by ArchieLuxton** [comment]
Thanks so much for your support @donmccurdy, a combination of all your snippets now gets precisely the result I was looking for.
